tests/tests/test-xcm/test-mock-dmp.ts

```diff
@@ -78,3 +79,153 @@ describeDevMoonbeam("Mock XCM - receive downward transfer", (context) => {
+    let weightPerMessage = (totalDmpWeight * BigInt(2)) / BigInt(numMsgs);
+
+    let weightPerXcmInst = 100_000_000n;
+    // Now we need to construct the message. This needs to:
+    // - pass barrier (withdraw + clearOrigin*n buyExecution)
+    // - fail in buyExecution, so that the previous instruction weights are counted
+    // we know at least 2 instructions are needed per message (withdrawAsset + buyExecution)
+    // how many clearOrigins do we need to append?
+
+    // we will bias this number. The reason is we want to test the decay, and therefore we need
+    // an unbalanced number of messages executed. We specifically need that at some point
+    // we get out of the loop of the execution (we reach the threshold limit), to then
+    // go on idle
+
+    // for that reason, we multiply times 2
+    let clearOriginsPerMessage = (weightPerMessage - weightPerXcmInst * 2n) / weightPerXcmInst;
+
+    let instructions = [];
+    instructions.push({
+      WithdrawAsset: [
+        {
+          id: {
+            Concrete: {
+              parents: 0,
+              interior: {
+                X1: { PalletInstance: balancesPalletIndex },
+              },
+            },
+          },
+          fun: { Fungible: 1 },
+        },
+      ],
+    });
+
+    // we push clear origins
+    for (let i = 0; i < clearOriginsPerMessage; i++) {
+      instructions.push({
+        ClearOrigin: null,
+      });
+    }
+
+    // we push the last buyExecution, that will fail
+    instructions.push({
+      BuyExecution: {
+        fees: {
+          id: {
+            Concrete: {
+              parents: 1,
+              interior: {
+                X2: [{ Parachain: ownParaId }, { PalletInstance: balancesPalletIndex }],
+              },
+            },
+          },
+          fun: { Fungible: 1 },
+        },
+        weightLimit: { Limited: new BN(20000000000) },
+      },
+    });
+
+    let xcmMessage = {
+      V2: instructions,
+    };
+
+    const receivedMessage: XcmVersionedXcm = context.polkadotApi.createType(
+      "XcmVersionedXcm",
+      xcmMessage
+    ) as any;
+
+    const totalMessage = [...receivedMessage.toU8a()];
+
+    // We want these isntructions to fail in BuyExecution. That means
+    // WithdrawAsset needs to work. The only way for this to work
+    // is to fund each sovereign account
+    const sovereignAddress = u8aToHex(
+      new Uint8Array([...new TextEncoder().encode("Parent")])
+    ).padEnd(42, "0");
+
+    // We first fund the parent sovereign account with 1000
+    // we will only withdraw 1, so no problem on this
+    await context.createBlock(context.polkadotApi.tx.balances.transfer(sovereignAddress, 1000n));
+
+    // now we start injecting messages
+    // several
+    for (let i = 0; i < numMsgs; i++) {
+      await customWeb3Request(context.web3, "xcm_injectDownwardMessage", [totalMessage]);
+    }
+
+    await context.createBlock();
```

Comment:
optionally to leverage `context.createBlock(<Array>)` signature:

```suggestion
    for (let i = 0; i < numMsgs; i++) {
      await customWeb3Request(context.web3, "xcm_injectDownwardMessage", [totalMessage]);
    }

    await context.createBlock(Array(numMsgs).fill(customWeb3Request(context.web3, "xcm_injectDownwardMessage", [totalMessage])));
```